### ISSUE TITLE
Autovalue not run for fields in an array of embedded documents when specified with a positional ("$") operator

### ISSUE THREAD

**Issue by fongandrew** [aldeed/meteor-simple-schema#222]
Given a document with a field specified as an array of embedded documents, like so:

```javascript
SubSchema = new SimpleSchema({
  value: {
    type: String,
    autoValue: function() {
      return "autovalue";
    }
  }
});

TestSchema = new SimpleSchema({
  children: {
    type: [SubSchema]
  }
});
```

A modifier that updates a value within a single embedded document specified with a position operator ('$') will not trigger the autovalue. That is, the following doesn't call the autovalue function:

```javascript
var mod = {
  $set: {"children.$.value": "should be overridden by autovalue"}
};
TestSchema.clean(mod, {isModifier: true});
```

Complete repro is at https://github.com/fongandrew/meteor-simple-schema-array-test

**Comment by Miges** [comment]
Having this issue too.  Works on insert but not afterwards.

**Comment by Firfi** [comment]
Can you try to rename 'children' to 'chldrn' or something else? Actually, I had another issue that field name and renaming fixed it.

**Comment by guidouil** [comment]
I have the same issue, it work as an insert not as an update. Here is the piece of the sub object in schema for a uniqueID to identify each sub object
```javascript
"answers.$.qaId": {
  type: String,
  autoValue: function() {
    if (!this.isSet) {
      return Random.id();
    }
  },
  autoform: {
    label: false,
    afFieldInput: {
      type: "hidden"
    }
  }
},
```
gonna try with a sub-schema

**Comment by fongandrew** [comment]
@Firfi Renaming the field doesn't fix the issue for me.

@guidouil Yeah, an actual sub-schema doesn't work for me, but as a work-around, you can always validate the sub-document separately and then insert into the main schema as a blackbox.

**Comment by guidouil** [comment]
@fongandrew if I put it in a sub schema and with a blackbox in the main schema I don't get my ID set in the sub object. How will you fix that ? Can I watch for this collection server side and add an ID in this subdoc when updated ? (since this does work on insert)

**Comment by fongandrew** [comment]
I mean, you just manually clean the object prior to updating yourself. Like so: 

```javascript
var newSubObjectAttrs = { ... };
subObjectSchema.clean({$set: newSubObjectAttrs}, {isModifier: true});

var newSubObjectAttrs2 = {};
for (var key in newSubObjectAttrs) {
  newSubObjectAttrs2["subObjects.$." key] = newSubObjectAttrs[key];
}

ParentObjectCollection.update({"subObjects._id": someId}, {$set: newAttrs2});
```

Not ideal to have to do this before every update, but that's why it's a workaround.

**Comment by guidouil** [comment]
@fongandrew sorry to ask again, but do you do that ? I'm using autoform, so it should be inside a hook I guess.

**Comment by fongandrew** [comment]
@guidouil I just use my own custom function in lieu of Collection.update to do it. An autoform hook works as well I guess, although I haven't played much with autoform. You could also try replacing the Collection.update function yourself if you're comfortable with that.

**Comment by guidouil** [comment]
thanks @fongandrew I did my workaround in a hook, also it doesn't use clean since I did not figure out how to use it (I certainly did not try hard enough). Here is my workaround if someone needs something like a unique ID in a sub document as I do :)
```javascript
var hooksObject = {
  'before': {
    // To add a unique _id in questions.aswers subobject when adding a new answer in existing question
    update: function(doc) {
      var formSet = doc.$set;
      var formUnset = doc.$unset;
      for (var i = 0; i < 100; i++) {
        if (formSet['answers.'+i+'.answer'] && !formSet['answers.'+i+'._id']) {
          formSet['answers.'+i+'._id'] = Random.id();
          delete formUnset['answers.'+i+'._id'];
        }
      }
      doc.$set = formSet;
      doc.$unset = formUnset;
      return doc;
    }
  }
};

AutoForm.hooks({
  questionUpdate: hooksObject
});
```

**Comment by willrbc** [comment]
I'm having a similar problem with A subschema inside an array.

````javascript
Schemas.Contacts.Base = new SimpleSchema({
    firstName: {
        type: String,
        max: 100
    },
    lastName: {
        type: String,
        max: 100
    },
    emails: {
        type: [Schemas.Contacts.Email],
        minCount: 1
    },
    initials: {
        type: String,
        autoValue: function() {
            console.log('initials');
            var firstName = this.siblingField("firstName");
            var lastName = this.siblingField("lastName")
            var initials = '';
            console.log(this.siblingField("firstName").value);
            if (firstName.isSet) {
                initials = firstName.value.charAt(0);
            };
            if (lastName.isSet) {
                initials = initials + lastName.value.charAt(0);
            };
            if (initials === '') {
                return this.unset();
            } else {
                return initials;
            }
        }
    }
});

````

The above schema sites inside a 'Company' schema that has many contacts,  I'd like to use the Initials as a short way.  Its not entirely necessary as I could just get the initials via a method in the view, but there is likely to be many other things I want to use autoValue for so it would be useful to have this working

**Comment by em0ney** [comment]
Having the same issue here.  Would be best if we could get collection2 to trigger the collection of autoValues for subschemas.

As pointed out above, it works on insert and in my case it works on $addToSet

@aldeed - any updates mate?

**Comment by em0ney** [comment]
@aldeed - I created a PR to fix this issue.  Please review.

Something else to notice for all those who may be inspecting sibling values in their autoValue hooks for subschemas/subdocuments:

You may need to use `this.siblingField('subDocField')` when using $addToSet or $push and `this.field('subDocField')` when using $set.

I believe this has something to do with `document.$.subDocField` being interpreted as being at the base level of the object by `this.siblingField`.

No matter, this has a simple work around:
```javascript
someAutoValueField: {
  type: Date,
  autoform: {
    omit: true
  },
  autoValue: function() {
    var subDocField = this.operator === '$set' ? this.field('document.$.subDocField') : this.siblingField('subDocField');
  }
}
```

**Comment by steve-ross** [comment]
I have a weird issue where my autovalue is being set properly on both the server and the client but never saved in the database... logging on both sides shows the proper value but, it never actually makes it to the database.

**Comment by steve-ross** [comment]
I figured out my issue... annoying that it isn't obvious but, since autoValue ALWAYS runs on the server (and the server wins)... so my code should have been:

```javascript
    autoValue: function () {
      // update deadline whenever departure is set
      // automatically set it based on departure if they didn't enter a date here
      if(!moment(new Date(this.value)).isValid()) this.unset();

      if(this.field('departs').isSet && !this.isSet) {
        var departs = this.field('departs').value;
        var deadline =  moment(departs).subtract(30, 'days').toDate();
        console.log('deadline', deadline);
        return deadline;
      }else{
        console.log('autovalue', this.value);
        return this.value;
      }

    }

```

**Comment by tylerhartley** [comment]
Any progress towards accepting the PR?

**Comment by jstoeffler** [comment]
> Any progress towards accepting the PR?

+1

**Comment by willrbc** [comment]
+2

I'm experimenting with Astronomy in parallel with Simple Schema and it appears both packages have this issue with automatic values / event function in subschemas

**Comment by ahmedtabrez** [comment]
:+1: 

`autoValue` doesn't work when using `$push` modifier to update an array field.

**Comment by keyscores** [comment]
+1 seems like the pull request has passing tests. Any reason not to merge it? It will soon be one year since the pull request was made.

**Comment by keyscores** [comment]
Just to be clear to future readers. AutoValue's this.isSet will always be false when using in a subdocument array.
This document structure will be a problem for Autovalue:

```
new SimpleSchema({
  user: {
   type: Array,
  },
  'user.$': {
     type: Object
  },
  'user.$.fullname': { 
   autoValue: function(){
     console.log('isSet', this.isSet  );
     //always returns false
    }
  }
});
```

**Comment by EddyBorja** [comment]
Just ran into this issue, autoValue not working with $push. Pull request should probably get merged, it'll be a year soon!

**Comment by em0ney** [comment]
@aldeed - any reason why this hasn't been merged?

**Comment by cdolek** [comment]
@aldeed is there any progress about this issue?

**Comment by keyscores** [comment]
@aldeed, I know you're probably overloaded with other work. Are there any other contributors with github permissions to merge passing pull requests?

**Comment by thebarty** [comment]
Yeah, I just ran into this issue... and damn I've lost some hours on this one... 
PLUS ONE

**Comment by thebarty** [comment]
Hi guys,

I spent some time trying to get this to work, but I failed - just letting you know about the progress.

1) This topic is related to this issue, where ``custom()``-functions are NOT called on ``$set: ``-commands. So when fixing this, ``custom()`` functions should also be included? https://github.com/aldeed/meteor-collection2/issues/318

2) I tried getting the pull request to work, but it failed for me. Read about my experience in my comment here: https://github.com/aldeed/meteor-simple-schema/pull/411.

By the way - this is my example use-case:
```
MySchema = new SimpleSchema({
  accounts: {
    type: [Object],
    optional: true,
  },
  'accounts.$.password': {
    type: String,
    label: 'Enter a password',
  },
  'accounts.$.confirmPassword': {
    type: String,
    label: 'Enter the password again',
    autoValue: function () {
      // BUG??? THIS IS NOT CALLED, when doing a ``$set: ``
      console.log('autoRun')
    }
  }
})
Collection = new Mongo.Collection('myschema')
Collection.attachSchema(MySchema)

Collection.remove({})
var id = Collection.insert({
  accounts: [
    {password: '12345678', confirmPassword: '12345678'}, 
    {password: '23456789', confirmPassword: '23456789'}
  ]
})

// BUG???
// THIS CALL GOES THRU WITHOUT "autoValue() of "accounts.$.confirmPassword" being called
Collection.update(id, {$set: {
  'accounts.1.password': '9999999999999'
}})
```

So I am still stuck trying to find a workaround.

**Comment by thebarty** [comment]
@keyscores: Have you found a solution for this?

by the way - I found that ``user.autoValue`` is called... see here
```
new SimpleSchema({
  user: {
   type: Array,
   autoValue: function() {
     // this function is triggered correctly, but it does not really help, does it?
     console.log('this is called on an $set : {}')
   }
  },
  'user.$': {
   type: Object
  },
  'user.$.fullname': { 
   autoValue: function(){
     console.log('isSet', this.isSet  );
     //always returns false
    }
  }
});
```

**Comment by keyscores** [comment]
@thebarty the autovalue does not trigger for any sub schemas . In the example above, yes the top level collection.User will trigger autovalue. But it's not helpful, since, it may return isSet = true, when really you need to check if user.$.fullname.isSet = true. The work-around for now, temporarily, is to use an Autoform hook to process the data on insert and update. It's not as convenient as simpleschema autovalue, but it will likely give you what you need. https://github.com/aldeed/meteor-autoform#callbackshooks

**Comment by aldeed** [comment]
Thanks for your patience. SimpleSchema 2.0.0-rc.1 is now released and should fix this bug.

There are a number of breaking changes when updating to 2.0, so be sure to check out [the change log](https://github.com/aldeed/meteor-simple-schema/blob/devel/CHANGELOG.md#200). If you use aldeed:collection2, you will need to use 2.10.0 or higher of that package in order to use SimpleSchema 2.0. If you use autoform, it is not yet updated to work with SimpleSchema 2.0, but hopefully soon.

SimpleSchema is now an isomorphic NPM package, so you can check out the updated readme and file issues over at the [other repo](https://github.com/aldeed/node-simple-schema). The Meteor wrapper package will exist for now but eventually I will probably deprecate support for it.

This is still a beta/RC and I do expect people will find issues, so use with caution. Production use is not yet recommended. That said, there are more and better unit tests than ever before, and the codebase should be much easier for others to read through and debug quickly.

**Comment by wirelineio** [comment]
@aldeed  thanks again and great job, on 2.0!

**Comment by micha1983** [comment]
Has somebody tested this. for me it's still not working.
